alf/utils/process_coordinator_test.py

```diff
@@ -0,0 +1,432 @@
+
+
+def stop_on_event(coord, wait_for_stop, set_when_stopped):
+    wait_for_stop.wait()
+    coord.request_stop()
+    set_when_stopped.set()
+
+
+def raise_on_event(coord, wait_for_stop, set_when_stopped, ex,
+                   report_exception):
+    try:
+        wait_for_stop.wait()
+        raise ex
+    except RuntimeError as e:
+        if report_exception:
+            coord.request_stop(e)
+        else:
+            coord.request_stop(sys.exc_info())
+    finally:
+        if set_when_stopped:
+            set_when_stopped.set()
+
+
+def raise_on_event_using_context_handler(coord, wait_for_stop,
+                                         set_when_stopped, ex):
+    with coord.stop_on_exception():
+        wait_for_stop.wait()
+        raise ex
+    if set_when_stopped:
+        set_when_stopped.set()
+
+
+def sleep_a_bit(n_secs):
+    time.sleep(n_secs)
+
+
+def wait_for_processes_to_register(coord, num_processes):
+    while True:
+        with coord._lock:
+            if len(coord._registered_processes) == num_processes:
+                break
+        time.sleep(0.001)
+
+
+class CoordinatorTest(test.TestCase):
+    def testStopAPI(self):
+        coord = coordinator.Coordinator()
+        self.assertFalse(coord.should_stop())
+        self.assertFalse(coord.wait_for_stop(0.01))
+        coord.request_stop()
+        self.assertTrue(coord.should_stop())
+        self.assertTrue(coord.wait_for_stop(0.01))
+
+    def testStopAsync(self):
+        coord = coordinator.Coordinator()
+        self.assertFalse(coord.should_stop())
+        self.assertFalse(coord.wait_for_stop(0.1))
+        wait_for_stop_ev = Event()
+        has_stopped_ev = Event()
+        t = Process(
+            target=stop_on_event,
+            args=(coord, wait_for_stop_ev, has_stopped_ev))
+        t.start()
+        self.assertFalse(coord.should_stop())
+        self.assertFalse(coord.wait_for_stop(0.01))
+        wait_for_stop_ev.set()
+        has_stopped_ev.wait()
+        self.assertTrue(coord.wait_for_stop(0.05))
+        self.assertTrue(coord.should_stop())
+
+    def testJoin(self):
+        coord = coordinator.Coordinator()
+        processes = [
+            Process(target=sleep_a_bit, args=(0.01, )),
+            Process(target=sleep_a_bit, args=(0.02, )),
+            Process(target=sleep_a_bit, args=(0.01, ))
+        ]
+        for t in processes:
+            t.start()
+        coord.join(processes)
+        for t in processes:
+            self.assertFalse(t.is_alive())
+
+    def testJoinAllRegistered(self):
+        coord = coordinator.Coordinator()
+        processes = [
+            Process(target=sleep_a_bit, args=(0.01, )),
+            Process(target=sleep_a_bit, args=(0.02, )),
+            Process(target=sleep_a_bit, args=(0.01, ))
+        ]
+        for t in processes:
+            t.start()
+        for p in processes:
+            coord.register_process(p)
+        wait_for_processes_to_register(coord, len(processes))
```

Comment:
Looks like this wait_for_processes_to_register won't be called automatically after a process is registered, which is dangerous. Should it be included in `register_process`?

Reply:
Yeah I think it's better to let the main process to do all the registrations.

Reply:
Passing coord into subprocess init will register, right? The init function is still inside the main process until you call .start().

Reply:
Oh, yes, you are right.  Do you also want me to use MyProcess.__init__ to register these simple test processes, or are you Ok with the current code (manually registering)?

I have other test code using MyProcess lower down in the file, which does registration in __init__.

Reply:
I think for testing code, either way is fine.